Arc: CDI Decorators and interface methods with default implementations

### Describe the bug

Found a tricky bug while using CDI Decorators based on an interface that has methods with default implementations.

Example:

```java
interface Interface {
  default int getWithDefault() { return 3; }
}

@Decorator
class InterfaceDecorator implements Interface {
  @Inject
  @Delegate
  Interface delegate;

  public int getWithDefault() {
    return delegate.getWithDefault();
  }
}

@Dependent
class ConcreteImplementation implements Interface {
  @Override
  public int getWithDefault() {
    return 4;
  }
}

class User {
  @Inject
  Interface anInterface;

  public void call() {
    assert: anInterface.getWithDefault() == 4; // value of concrete implementation
  }
}
```

This code will end in an AssertionError cause the concerte implementation never gets called.

### Expected behavior

Interface methods with default implementation are still override by the container to forwarding calls from one decorator to another till the concrete implementation of the type.

### Actual behavior

Calls to interface methods with default implementation are not forwarded by the container from the first decorator to the second decorator or the actual implementation.

### How to Reproduce?

Reproducer: 
[reproducer-decorator-default-impl.zip](https://github.com/quarkusio/quarkus/files/12488235/reproducer-decorator-default-impl.zip)

Run `mvn clean install` after unzip.

### Output of `uname -a` or `ver`

6.2.0 Ubuntu

### Output of `java -version`

20.0.2

### GraalVM version (if different from Java)

23.0.1

### Quarkus version or git rev

3.3.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.4

### Additional information

_No response_

The default methods are definitely removed after this contribution:

https://github.com/quarkusio/quarkus/pull/33841/files#diff-f05a30ce973915337c900f71476c739f710d3fcc1ab6d148bbb418755a2bd5a9

But when I try to remove the condition, obviously few tests start failing again:

```
[ERROR] Errors: 
[ERROR]   MultipleDecoratorsWithTypeVariablesTest.testConsumer » NullPointer
[ERROR]   MultipleDecoratorsWithTypeVariablesTest.testFunction » NullPointer
[ERROR]   MultipleDecoratorsWithTypeVariablesTest.testSupplier » NullPointer
[ERROR]   SlightlyMoreComplexDecoratorWithTypeVariableTest.testFunction » NullPointer
```

I don't think I am in a position to propose a fix.

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Methods.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Methods.java
@@ -140,7 +140,7 @@ final class Methods {
     }
 
     static boolean skipForDelegateSubclass(MethodInfo method) {
-        if (Modifier.isStatic(method.flags()) || method.isSynthetic() || isDefault(method)) {
+        if (Modifier.isStatic(method.flags()) || method.isSynthetic()) {
             return true;
         }
         if (IGNORED_METHODS.contains(method.name())) {

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java
@@ -637,13 +637,10 @@ public class SubclassGenerator extends AbstractGenerator {
         }
 
         List<String> constructorParameterTypes = new ArrayList<>();
-        // Fields and constructor
-        FieldCreator subclassField = null;
-        if (decoratedMethods.size() != nextDecoratorsValues.size()) {
-            subclassField = delegateSubclass.getFieldCreator("subclass", subclass.getClassName())
-                    .setModifiers(ACC_PRIVATE | ACC_FINAL);
-            constructorParameterTypes.add(subclass.getClassName());
-        }
+        // Holds a reference to the subclass of the decorated bean
+        FieldCreator subclassField = delegateSubclass.getFieldCreator("subclass", subclass.getClassName())
+                .setModifiers(ACC_PRIVATE | ACC_FINAL);
+        constructorParameterTypes.add(subclass.getClassName());
         Map<DecoratorInfo, FieldDescriptor> nextDecoratorToField = new HashMap<>();
         for (DecoratorInfo nextDecorator : decoratorParameters) {
             FieldCreator nextDecoratorField = delegateSubclass
@@ -653,6 +650,7 @@ public class SubclassGenerator extends AbstractGenerator {
             nextDecoratorToField.put(nextDecorator, nextDecoratorField.getFieldDescriptor());
         }
 
+        // Constructor
         MethodCreator constructor = delegateSubclass.getMethodCreator(Methods.INIT, "V",
                 constructorParameterTypes.toArray(new String[0]));
         int param = 0;
@@ -664,10 +662,8 @@ public class SubclassGenerator extends AbstractGenerator {
                     constructor.getThis());
         }
         // Set fields
-        if (subclassField != null) {
-            constructor.writeInstanceField(
-                    subclassField.getFieldDescriptor(), constructor.getThis(), constructor.getMethodParam(param++));
-        }
+        constructor.writeInstanceField(
+                subclassField.getFieldDescriptor(), constructor.getThis(), constructor.getMethodParam(param++));
         for (FieldDescriptor field : nextDecoratorToField.values()) {
             constructor.writeInstanceField(
                     field, constructor.getThis(), constructor.getMethodParam(param++));
@@ -700,10 +696,6 @@ public class SubclassGenerator extends AbstractGenerator {
 
         for (MethodKey m : methods) {
             MethodInfo method = m.method;
-            if (Methods.skipForDelegateSubclass(method)) {
-                continue;
-            }
-
             MethodDescriptor methodDescriptor = MethodDescriptor.of(method);
             MethodCreator forward = delegateSubclass.getMethodCreator(methodDescriptor);
             // Exceptions
@@ -818,9 +810,7 @@ public class SubclassGenerator extends AbstractGenerator {
         // Create new delegate subclass instance and set the DecoratorDelegateProvider to satisfy the delegate IP
         ResultHandle[] paramHandles = new ResultHandle[constructorParameterTypes.size()];
         int paramIdx = 0;
-        if (subclassField != null) {
-            paramHandles[paramIdx++] = subclassConstructor.getThis();
-        }
+        paramHandles[paramIdx++] = subclassConstructor.getThis();
         for (DecoratorInfo decoratorParameter : decoratorParameters) {
             ResultHandle decoratorHandle = decoratorToResultHandle.get(decoratorParameter.getIdentifier());
             if (decoratorHandle == null) {

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/decorators/DecoratorDefaultMethodTest.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/decorators/DecoratorDefaultMethodTest.java
@@ -1,0 +1,59 @@
+package io.quarkus.arc.test.decorators;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import jakarta.annotation.Priority;
+import jakarta.decorator.Decorator;
+import jakarta.decorator.Delegate;
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.enterprise.util.TypeLiteral;
+import jakarta.inject.Inject;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.test.ArcTestContainer;
+
+public class DecoratorDefaultMethodTest {
+    @RegisterExtension
+    public ArcTestContainer container = new ArcTestContainer(Converter.class, ToLengthConverter.class,
+            NoopConverterDecorator.class);
+
+    @SuppressWarnings("serial")
+    @Test
+    public void testDecoration() {
+        Converter<String> converter = Arc.container().instance(new TypeLiteral<Converter<String>>() {
+        }).get();
+        assertEquals(5, converter.convert("Hola!"));
+    }
+
+    interface Converter<T> {
+        default int convert(T value) {
+            return Integer.MAX_VALUE;
+        }
+    }
+
+    @ApplicationScoped
+    static class ToLengthConverter implements Converter<String> {
+        @Override
+        public int convert(String value) {
+            return value.length();
+        }
+    }
+
+    @Priority(1)
+    @Decorator
+    static class NoopConverterDecorator implements Converter<String> {
+
+        @Inject
+        @Delegate
+        Converter<String> delegate;
+
+        @Override
+        public int convert(String value) {
+            return delegate.convert(value);
+        }
+    }
+
+}

